### PR TITLE
Reduce payout confirmation wait times

### LINKED ISSUE #353: 'Confirm' button takes too much time (10-25 mins) to be enabled
### Description

*Type*: Bug

*Summary*: after organization merge PR, event 'Won ... (payout pending maintainer confirmation)' immediately appears in Activity feed. But in order to `Confirm` payment user should wait 10-25 min.
During this time issue can appear in 'Bounties' list if it wasn't closed automatically by Github or in case it was reopened.
confirm
 

#### Expected behavior
7) issue doesn't appear on 'Bounties' list anymore even it was reopened
8) 'Confirm' button is available after merging PR (2-5 minutes);
#### Actual behavior
7) if you reopen issue it appears in 'Bounties' list
8) 'Confirm' button is available after 10-20 mins;

### Reproduction

*Requirements*: GH organization account is whitelisted, signed app, test application is added to repo;
1)  create issue 
2)  add `bounty` label, fund contract
3)  create PR to fix it as developer 
4)  merge PR
5) navigate to `Activity` in https://testing.openbounty.status.im/app. 
`Won ....` should be displayed right after merging PR.
6) reopen issue
7) navigate to `Bounties` in https://testing.openbounty.status.im/app. 
8) log in as organization and wait until 'Confirm' button will be enabled

## Additional Information
* Browser: Chrome 64
* Operating System: Mac OS High Sierra

### PR DESCRIPTION
**Pull request by siphiuel**
Fixes #353 

This fix does the following:
  - Move balances updates to 10-min scheduler threads
  - Cache return values of `eth.core/create-web3j` and `eth.core/creds` in an atom, reset these when mounting corresponding state
  - Add Tufte macros for profiling, can be removed afterwards

### CODE REVIEW

File: src/clj/commiteth/eth/core.clj [hunk 1]
```diff
@@ -26,24 +28,30 @@
...
+  (or @creds-obj
+      (let [password  (wallet-password)
+            file-path (wallet-file-path)]
+        (if (and password file-path)
+          (swap! creds-obj
+                 (constantly (WalletUtils/loadCredentials
+                               password
+                               file-path)))
+          (throw (ex-info "Make sure you provided proper credentials in appropriate resources/config.edn"
+                          {:password password :file-path file-path}))))))
```
Comment (martinklepsch) [inline src/clj/commiteth/eth/core.clj]: memoize or defonce might be interesting alternatives to the atom approach here..

**Comment by siphiuel** [comment]
Some profiling data is in order. As it turns out, `update-contract-internal-balances`, `update-open-issue-usd-values`, `update-balances` are the slowest (run on test DB):
```
:update-issue-contract-address 11.74ms
:deploy-pending-contracts 13.09ms
:self-sign-bounty 1.81s
:update-confirm-hash 13.51ms
:update-watch-hash 10.89ms
:update-payout-receipt 3.68s
:update-contract-internal-balances 2.79m
:update-open-issue-usd-values 5.25m
:update-balances 4.3m
```
Changing `create-web3j` and `creds` to cache their return values in an atom helped a bit.
Using the following code
```clj
(profile {}
           (doseq [i (range 5)]
             (update-contract-internal-balances) 
             (update-open-issue-usd-values) 
             (update-balances)))
  ```
Before:
```
                               pId      nCalls       Min        Max       MAD      Mean   Time% Time
                   :token-balances       1,320     1.15s      2.56s  118.75ms     1.35s      49 29.68m
     :update-open-issue-usd-values           5     5.19m      5.31m     1.78s     5.25m      44 26.25m
                  :update-balances           5     3.99m      4.18m     3.17s     4.07m      34 20.35m
                     :getTokenList       1,320  580.29ms      1.91s   77.94ms  705.77ms      26 15.53m
:update-contract-internal-balances           5     2.56m       2.8m      5.4s      2.7m      22 13.5m
             :load-bounty-contract       1,320  548.51ms   785.69ms   34.67ms  605.19ms      22 13.31m
                    :token-balance       2,180  290.39ms      1.63s   54.99ms  352.88ms      21 12.82m
                  :get-balance-eth       1,320  289.72ms      1.32s   55.98ms  358.56ms      13 7.89m
          :token-balance-in-bounty          70     1.15s      1.46s   71.64ms      1.3s       3 1.51m
                         :getValue       1,320    7.66μs   195.72μs    3.01μs   14.88μs       0 19.64ms
                        Clock Time                                                          100 60.11m
                    Accounted Time                                                          234 140.85m
```
After:
```
                               pId      nCalls       Min        Max       MAD      Mean   Time% Time
     :update-open-issue-usd-values           5      3.1m      3.37m     5.56s      3.2m      41 16.01m
                    :token-balance       2,180  290.84ms   950.08ms   52.55ms  360.35ms      34 13.09m
:update-contract-internal-balances           5     2.47m      2.59m     2.43s     2.54m      33 12.69m
                  :update-balances           5     1.92m      2.06m     2.69s     1.99m      26 9.93m
                   :token-balances       1,320  290.91ms      1.35s   88.82ms  392.41ms      22 8.63m
                  :get-balance-eth       1,320  289.94ms      1.03s   55.32ms   357.0ms      20 7.85m
                     :getTokenList       1,320  290.83ms      1.35s   48.47ms  354.21ms      20 7.79m
          :token-balance-in-bounty          70  292.56ms   870.37ms   61.01ms  375.36ms       1 26.28s
             :load-bounty-contract       1,320   29.06μs   206.26μs    4.76μs   43.98μs       0 58.05ms
                         :getValue       1,320    6.97μs    50.33μs    1.21μs   11.07μs       0 14.62ms
                        Clock Time                                                          100 38.63m
                    Accounted Time                                                          198 76.44m
```

File: src/clj/commiteth/eth/core.clj (hunk 1 above)
Comment (siphiuel) [reply src/clj/commiteth/eth/core.clj]: Yes indeed! For instance, `memo` together with `memo-clear!` from [core.memoize](https://github.com/clojure/core.memoize). We can sprinkle it around the code for some low-hanging fruit.

Comment (martinklepsch) [reply src/clj/commiteth/eth/core.clj]: I was actually thinking of Clojure.core functions memoize and defonce. Another alternative might be treating these objects as 'mount' components

**Comment by churik** [comment]
### Environments
* Ethereum wallet used: MetaMask
* Browser: Chrome, Firefox 58, Safari 11
* Operating System: macOS High Sierra 10.13.3

### Tested functionality
* Creating new issue with bounty;
* Signing up new repo for bounty issues;
* Confirming bounty payments 

Payout confirmation button is enabled during ~1 min for now.
Thanks @siphiuel , amazing work!

**Comment by martinklepsch** [comment]
Just so I understand: previously it took 10-25 min for the confirm button to be enabled & now it should take about 1 minute? 

What are the remaining obstacles to make it instantaneous?

**Comment by churik** [comment]
> Just so I understand: previously it took 10-25 min for the confirm button to be enabled & now it should take about 1 minute? 

Yes, it is!
Also really annoying that it is not updated automatically, so you should constantly reload page in order to see button enabled

**Comment by siphiuel** [comment]
@martinklepsch Cannot be instantaneous unfortunately, as we need to mine confirmation transaction and wait for the callback we pass to be executed (relevant code is in `:confirm-payout` event handler in`handlers.cljs`).
@churik It's not updated automatically as the relevant event handler (`:owner-bounties`) is not executed via JS timer each minute. However, @martinklepsch suggested earlier that we can make `:owner-bounties` a child of `:open-bounties`, and they are fetched each minute, so this should fix the problem.

**Comment by churik** [comment]
@siphiuel create separate issue for this fix? or you can say in what PR it will be fixed?

**Comment by siphiuel** [comment]
Please create a separate issue I guess.

**Comment by churik** [comment]
Created #361

**Comment by martinklepsch** [comment]
> Cannot be instantaneous unfortunately, as we need to mine confirmation transaction and wait for the callback we pass to be executed (relevant code is in `:confirm-payout` event handler in `handlers.cljs`).

Got it. In this case we should consider looking for other ways to make this smoother. Perhaps push that information to the user using the Status app or a similar "push mechanism". This is further down the road obviously but it's probably good for us (maybe @EugeOrtiz in particular) to keep in mind.

**Pull request merged**